experimental/sql/src/alg_to_ssa.py

```diff
@@ -0,0 +1,151 @@
+
+import dialects.rel_alg as RelAlg
+import dialects.rel_ssa as RelSSA
+
+# This file contains the rewrite infrastructure to translate the relational
+# algebra dialect to the relational SSA dialect. The current design has a parent
+# class `RelAlgRewriter` that contains functions used for several `Rewriter`s.
+# All other `Rewriter`s inherit from that class. To introduce SSA, the passes
+# insert the children, that should become operands, before the matched op, then
+# use the `added_operations_before` functionality of the rewriter to link the
+# operands to the new operation.
+
+
+@dataclass
+class RelAlgRewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelAlg.DataType) -> RelSSA.DataType:
+    if isinstance(type_, RelAlg.String):
+      return RelSSA.String.get(type_.nullable)
+    if isinstance(type_, RelAlg.Int32):
+      return RelSSA.Int32()
+    raise Exception(
+        f"datatype conversion not yet implemented for {type(type_)}")
+
+  def lookup_type_in_schema(self, name: str,
+                            bag: RelSSA.Bag) -> Optional[RelSSA.DataType]:
+    """
+    Looks up the type of name in the schema of bag.
+    """
+    for s in bag.schema.data:
+      if s.elt_name.data == name:
+        return s.elt_type
+    return None
+
+  def lookup_type_in_parent_op(self, name: str, parent_op: Operation):
+    """
+    Crawls through all parent_ops until reaching either a ModuleOp, in which
+    case the lookup failed or reaching an operation with an input bag, that the
+    type can be looked up in.
+    """
+    if isinstance(parent_op, ModuleOp):
+      raise Exception(f"element not found in parent schema: {name}")
+    if isinstance(parent_op, RelAlg.Select):
+      type_ = self.lookup_type_in_schema(name,
+                                         parent_op.input.op.results[0].typ)
+      if type_:
+        return type_
+      raise Exception(f"element not found in parent schema: {name}")
+    return self.lookup_type_in_parent_op(name, parent_op.parent_op())
+
+
+#===------------------------------------------------------------------------===#
+# Expressions
+#===------------------------------------------------------------------------===#
+"""
+All expression rewriters implicitely assume that the last operation in a lock is
+the one to be yielded.
+"""
+
+
+@dataclass
+class LiteralRewriter(RelAlgRewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Literal, rewriter: PatternRewriter):
+    new_op = RelSSA.Literal.get(op.val, self.convert_datatype(op.type))
+    rewriter.insert_op_before_matched_op(
+        RelSSA.Literal.get(op.val, self.convert_datatype(op.type)))
+    rewriter.erase_matched_op()
+
+
+@dataclass
+class ColumnRewriter(RelAlgRewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Column, rewriter: PatternRewriter):
+    res_type = self.lookup_type_in_parent_op(op.col_name.data, op.parent_op())
```

Comment:
You can actually call this function on `op` instead of `op.parent_op()` (and call the parameter of that function `op` instead of `parent_op`). It's a bit shorter, and the function name already implies that it will look at the parent...